### PR TITLE
Support for `triomphe::Arc`

### PR DESCRIPTION
**Pull request by michaelsproul**
This PR adds support for the alternative `Arc` implementation from [`triomphe`](https://docs.rs/triomphe/latest/triomphe/). I found that `triomphe` improved the memory efficiency and performance of a persistent data structure that I implemented, and was wondering if the same might apply to `rpds`. Spoiler: it does!

The code is mostly a copy-pasta of the existing `Arc` implementation.

### CODE REVIEW

**Comment by michaelsproul** [comment]
Some very rough benchmarks seem to show improvements of up to 40% for `rpds::List`!

Benchmark code here: https://github.com/orium/rpds/compare/master...michaelsproul:rpds:triomphe-bench?expand=0. Numbers from an Apple Silicon M1 Pro: https://gist.github.com/michaelsproul/2357776fd90632637108f60aa1738a51.

I'll try to run some more of the `rpds` benchmarks shortly.

I'll mark this PR as non-draft for now just to see if it passes CI (I have a compile-test failure locally that I haven't debugged yet).

**Comment by orium** [comment]
Thanks for this. I should be able to review this week.

**Comment by michaelsproul** [comment]
Awesome thanks @orium! In terms of interface do you prefer a separate ArcTK type, or a feature that swaps out the Arc impl. I think the first is probably better as it's more general, but I may use the second approach for the benchmarks to avoid code duplication and allow easy comparisons (just run the benches with/without the feature)

**Comment by orium** [comment]
> In terms of interface do you prefer a separate ArcTK type, or a feature that swaps out the Arc impl?

I might even prefer to not use the std `Arc` at all (i.e. no feature flag). `triomphe` is widely used, tested, and maintained. It also supports `no_std`, so we are good there. Do you see any reason to not make `ArcK` use `triomphe` unconditionally?

File: Cargo.toml [hunk 1]
```diff
@@ -41,6 +41,7 @@ codecov = { repository = "orium/archery", branch = "master", service = "github"
 
 [dependencies]
 static_assertions = "1.1.0"
+triomphe = { version = "0.1.9", optional = true }
```
Comment (orium) [inline Cargo.toml:44]: I think we also want no default features both to have no_std and to not have servo as a dependency.

**Comment by michaelsproul** [comment]
> Do you see any reason to not make ArcK use triomphe unconditionally?

I feel like for a generic lib like `archery` it's better to support both. Some users may have reason to prefer `std::sync::Arc`, e.g. if using it with `self: Arc<Self>`.

However I agree you'd be justified in switching out the default Arc backend for `rpds`. Providing up to 40% better performance without requiring users to fiddle with type parameters is a great win IMO.

File: Cargo.toml (hunk 1 above)
Comment (michaelsproul) [reply Cargo.toml]: Great idea. Done. No more `serde` or `stable_deref_trait` deps.

**Comment by orium** [comment]
Sounds good. Let's have a `ArcTK` behind a feature flag that is disabled by default.

**Comment by michaelsproul** [comment]
Awesome. I'll update the CI tests to run with triomphe enabled too

**Review by orium — COMMENTED** [review]
Looks good overall, but left a couple comments.

You also need to change the `tools/codecov.sh` script to call tarpaulin with `--all-features`, otherwise it won't run the `ArcT` tests (add it as the last flag to be consistent with my other projects: https://github.com/orium/rpds/commit/190a3eaf235221ace92214e10fc8cc657b4d3aee).

File: benches/archery_shared_pointer_arc.rs [hunk 2]
```diff
@@ -11,6 +11,9 @@ use std::ops::Deref;
 use criterion::{criterion_group, criterion_main, Criterion};
 use std::hint::black_box;
 
+#[cfg(feature = "triomphe")]
```
[the remark this answers is not in the document]
Comment (orium) [reply benches/archery_shared_pointer_arc.rs:14]: You forgot to remove these lines here. Also remove the trailing `,` in line 47.

[the remark this answers is not in the document]
Comment (michaelsproul) [reply benches/archery_shared_pointer_arc.rs]: Good catch, thanks! Sorry, just woke up. I think I've got everything now

**Comment by orium** [comment]
Merged. Thanks. I'll cut a release in the next few days.

**Pull request merged**